### PR TITLE
feat(template): added kotlin template for app-generator

### PR DESCRIPTION
**Pull request by sundardamon**
# Description

Closes : 

# Checklist:

- [x] My code follows the contribution guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My commits follow [conventional commit message guidelines](https://www.conventionalcommits.org/en/v1.0.0/)

### CODE REVIEW

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/pom.xml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.dfm</groupId>
+      <artifactId>rest-adapter</artifactId>
+      <exclusions>
+        <exclusion>
+          <artifactId>classmate</artifactId>
+          <groupId>com.fasterxml</groupId>
+        </exclusion>
+      </exclusions>
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/pom.xml:21]: this should be managed within main pom.xml

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/pom.xml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    </dependency>
+    <!-- Test -->
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-test</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.module</groupId>
+      <artifactId>jackson-module-kotlin</artifactId>
+    </dependency>
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/pom.xml:39]: Is this a test dependency ?

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/pom.xml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>com.github.cukedoctor</groupId>
+        <artifactId>cukedoctor-maven-plugin</artifactId>
+        <version>${cukedoctor-maven-plugin.version}</version>
+        <configuration>
+          <featuresDir>target/cucumber</featuresDir>
+          <outputFileName>ExamplePoetry</outputFileName>
+          <documentTitle>Example Poetry - Living Documentation</documentTitle>
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/pom.xml:50]: remove the word poetry

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    private val port: Int = 0
+    private var responseEntity: ResponseEntity<ExampleInfo>? = null
+
+    init {
+
+        DataTableType { row: Map<String, String> -> Example(1L, row["description"]) }
+        DataTableType { row: Map<String, String> -> ExampleEntity(1L, row["description"]) }
+
+        Before(HookNoArgsBody { poetryDao.deleteAll() })
+        After(HookNoArgsBody { poetryDao.deleteAll() })
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt:45]:
```
        Before(HookNoArgsBody { exampleDao.deleteAll() })
        After(HookNoArgsBody { exampleDao.deleteAll() })
```

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import packageName.domain.model.Example
+import packageName.domain.model.ExampleInfo
+import packageName.repository.dao.ExampleDao
+import packageName.repository.entity.ExampleEntity
+
+
+@ExtendWith(SpringExtension::class)
+@SpringBootTest(classes = [ExampleE2EApplication::class], webEnvironment = RANDOM_PORT)
+@ContextConfiguration(classes = [ExampleE2EApplication::class], loader = SpringBootContextLoader::class)
+class ExampleStepDef(restTemplate: TestRestTemplate, poetryDao: ExampleDao) : En {
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt:28]: rename `poetryDao` to `exampleDao`

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    companion object {
+        private const val LOCALHOST = "http://localhost:"
+        private const val API_URI = "/api/v1/examples"
+    }
+
+    @LocalServerPort
+    private val port: Int = 0
+    private var responseEntity: ResponseEntity<ExampleInfo>? = null
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt:37]: Use `lateinit`

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        When("user requests for all examples") {
+            val url = LOCALHOST + port + API_URI
+            responseEntity = restTemplate.getForEntity(url, ExampleInfo::class.java)
+        }
+
+        Then("the user gets the following examples") { dataTable: DataTable ->
+            val expectedExampleInfo = dataTable.asList<Example>(Example::class.java)
+            assertThat(responseEntity!!.statusCode).isEqualTo(HttpStatus.OK)
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt:59]: with `lateinit` of `responseEntity` you can avoid `!!`

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            poetryDao.saveAll(poems)
+        }
+
+        When("user requests for all examples") {
+            val url = LOCALHOST + port + API_URI
+            responseEntity = restTemplate.getForEntity(url, ExampleInfo::class.java)
+        }
+
+        Then("the user gets the following examples") { dataTable: DataTable ->
+            val expectedExampleInfo = dataTable.asList<Example>(Example::class.java)
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt:58]: rename `expectedExampleInfo` to `expectedExamples`

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            assertThat(responseEntity!!.statusCode).isEqualTo(HttpStatus.OK)
+            assertThat(responseEntity!!.body).isNotNull
+            assertThat(responseEntity!!.body!!.examples).isNotEmpty.extracting("description")
+                    .contains(*expectedExampleInfo.stream().map(Example::description).toArray())
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt:62]: do we really need to stream?

File: templates/hexagonal-spring-boot-kotlin/domain-api/src/main/kotlin/packageName/domain/model/Example.kt [hunk 10]
```diff
@@ -0,0 +1,3 @@
+package packageName.domain.model
+
+data class Example(val id: Long? = null, val description: String? = null)
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/domain-api/src/main/kotlin/packageName/domain/model/Example.kt:3]: `description` should not be null

File: templates/hexagonal-spring-boot-kotlin/jpa-adapter/pom.xml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-data-jpa</artifactId>
+      <version>${spring-boot.version}</version>
+      <exclusions>
+        <exclusion>
+          <artifactId>byte-buddy</artifactId>
+          <groupId>net.bytebuddy</groupId>
+        </exclusion>
+      </exclusions>
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/jpa-adapter/pom.xml:28]: exclusions are to be managed in dependencyManagement of main pom.xml

File: templates/hexagonal-spring-boot-kotlin/jpa-adapter/src/main/kotlin/packageName/repository/ExampleRepository.kt [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+import java.util.stream.Collectors
+import packageName.domain.model.Example
+import packageName.domain.port.ObtainExample
+import packageName.repository.dao.ExampleDao
+import packageName.repository.entity.ExampleEntity
+
+class ExampleRepository(private val exampleDao: ExampleDao) : ObtainExample {
+
+    override fun getAllExamples(): List<Example> {
+        return exampleDao.findAll().stream().map(ExampleEntity::toModel).collect(Collectors.toList());
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/jpa-adapter/src/main/kotlin/packageName/repository/ExampleRepository.kt:12]: do we really need to stream?

File: templates/hexagonal-spring-boot-kotlin/jpa-adapter/src/main/kotlin/packageName/repository/entity/ExampleEntity.kt [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+import javax.persistence.Table
+
+@Table(name = "T_EXAMPLE")
+@Entity
+data class ExampleEntity(
+        @Id
+        @GeneratedValue(strategy = AUTO)
+        private val id: Long? = null,
+        @Column(name = "DESCRIPTION")
+        private val description: String? = null) {
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/jpa-adapter/src/main/kotlin/packageName/repository/entity/ExampleEntity.kt:19]: should not be null

File: templates/hexagonal-spring-boot-kotlin/rest-adapter/pom.xml [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <parent>
+    <artifactId>artifactName</artifactId>
+    <groupId>org.dfm</groupId>
+    <version>1.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+  <artifactId>rest-adapter</artifactId>
+
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/rest-adapter/pom.xml:12]: avoid the extra space

File: templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        Before(HookNoArgsBody { poetryDao.deleteAll() })
+        After(HookNoArgsBody { poetryDao.deleteAll() })
+
+        Given("the following examples exists in the library") { dataTable: DataTable ->
+            val poems = dataTable.asList<ExampleEntity>(ExampleEntity::class.java)
+            poetryDao.saveAll(poems)
+        }
+
+        When("user requests for all examples") {
+            val url = LOCALHOST + port + API_URI
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/acceptance-test/src/test/kotlin/packageName/cucumber/ExampleStepDef.kt:53]:
```suggestion
            val url = $LOCALHOST$port$API_URI
```

File: templates/hexagonal-spring-boot-kotlin/rest-adapter/src/test/kotlin/packageName/rest/ExampleResourceTest.kt [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        assertThat(java.lang.Boolean.TRUE).isTrue()
+    }
+
+    @Test
+    fun `should give examples when asked for examples with the support of domain stub`() {
+        // Given
+        Mockito.lenient().`when`(requestExample!!.getExamples()).thenReturn(mockExampleInfo())
+        // When
+        val url = LOCALHOST + port + API_URI
+        val responseEntity = restTemplate!!.getForEntity(url, ExampleInfo::class.java)
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/rest-adapter/src/test/kotlin/packageName/rest/ExampleResourceTest.kt:47]:
```suggestion
            val url = $LOCALHOST$port$API_URI
```

File: templates/hexagonal-spring-boot-kotlin/rest-adapter/src/test/kotlin/packageName/rest/ExampleResourceTest.kt [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Test
+    fun `should start the rest adapter application`() {
+        assertThat(java.lang.Boolean.TRUE).isTrue()
+    }
+
+    @Test
+    fun `should give examples when asked for examples with the support of domain stub`() {
+        // Given
+        Mockito.lenient().`when`(requestExample!!.getExamples()).thenReturn(mockExampleInfo())
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/rest-adapter/src/test/kotlin/packageName/rest/ExampleResourceTest.kt:44]: avoid !!

File: templates/hexagonal-spring-boot-kotlin/rest-adapter/src/test/kotlin/packageName/rest/ExampleResourceTest.kt [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        // When
+        val url = LOCALHOST + port + API_URI
```
Comment (paul58914080) [inline templates/hexagonal-spring-boot-kotlin/rest-adapter/src/test/kotlin/packageName/rest/ExampleResourceTest.kt:46]:
```suggestion
            val url = $LOCALHOST$port$API_URI
```

File: templates/hexagonal-spring-boot-kotlin/jpa-adapter/pom.xml (hunk 11 above)
Comment (sundardamon) [reply templates/hexagonal-spring-boot-kotlin/jpa-adapter/pom.xml:28]: This exclusion belongs to the spring-boot-starter-data-jpa dependency which we are defining directly in the JPA adapter layer.

Comment (paul58914080) [reply templates/hexagonal-spring-boot-kotlin/jpa-adapter/pom.xml:28]: Manage the dependency like

**pom.xml**
https://github.com/devs-from-matrix/hexagonal-spring-boot-java/blob/master/pom.xml#L81-L91

**jpa-adapter/pom.xml**
https://github.com/devs-from-matrix/hexagonal-spring-boot-java/blob/master/jpa-adapter/pom.xml#L19-L22

**Comment by paul58914080** [comment]
Thank you @sundardamon for your work

**Pull request merged**